### PR TITLE
Add TLS v1.1 and TLS v1.2 support, disable TLS v1.0 by default on the…

### PR DESCRIPTION
**Pull request by laurynas-biveinis**
… server

This implements
https://blueprints.launchpad.net/percona-server/+spec/tls-1-2-5.5 (bug
1501089).

Backport from 5.7 a read-only, global-only variable tls-version that
lists the allowed TLS protocol versions for incoming connections: a
string with a comma-separated list of TLSv1, TLSv1.1, TLSv1.2. Set it to
TLSv1.1,TLSv1.2 by default. Do not backport the 5.7 MASTER_TLS_VERSION
clause of CHANGE MASTER TO statement.

For client side, only add support for TLSv1.1 and TLSv1.2 in addition to
the currently existing TLSv1.0, but do not add the corresponding
tls-version option as adding it requires adding MYSQL_OPT_TLS_VERSION to
libperconaserverclient and associated minor version bump, which is better
avoided.

Introduce new server read-only global variable have_tlsv1_2, which is set
to ON if the server has been compiled with a SSL library providing
TLSv1.2 support. Use this variable to skip TLSv1.2 tests on old
OpenSSL/YaSSL builds that do not support it.

At the same time fix bug 1582639 (ssl.cmake file is broken when using
custom OpenSSL build) to make it easier to build with OpenSSL in
non-default locations.

http://jenkins.percona.com/job/percona-server-5.5-param/1224/

### CODE REVIEW

File: vio/viosslfactories.c [hunk 1]
```diff
@@ -238,8 +386,14 @@ new_VioSSLFd(const char *key_file, const char *cert_file,
     NOTE: SSL_CTX_set_cipher_list will return 0 if
     none of the provided ciphers could be selected
   */
-  if (cipher &&
-      SSL_CTX_set_cipher_list(ssl_fd->ssl_context, cipher) == 0)
+  strcat(cipher_list, tls_cipher_blocked);
+  if (cipher)
+    strcat(cipher_list, cipher);
```
Comment (percona-ysorokin) [inline vio/viosslfactories.c:401]: possible `cipher_list` buffer overflow with malformed(too long) `cipher` parameter

**Comment by percona-ysorokin** [comment]
LGTM with buffer overflow comment addressed / impossibility of such case proven

File: vio/viosslfactories.c (hunk 1 above)
Comment (laurynas-biveinis) [inline vio/viosslfactories.c:401]: Good catch :)

# 

==19544==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff595e79e0 at pc 0x000108229555 bp 0x7fff595e69b0 sp 0x7fff595e6170
WRITE of size 4983 at 0x7fff595e79e0 thread T0
    #0 0x108229554 in wrap_strcat (libclang_rt.asan_osx_dynamic.dylib+0x41554)
    #1 0x107107b04 in new_VioSSLFd viosslfactories.c:391
    #2 0x1071081c4 in new_VioSSLAcceptorFd viosslfactories.c:504
    #3 0x10683a9c3 in init_ssl() mysqld.cc:3815
    #4 0x106836c8a in mysqld_main(int, char**) mysqld.cc:4775
    #5 0x7fff8e0d35ac in start (libdyld.dylib+0x35ac)
    #6 0x8  (<unknown module>)

Address 0x7fff595e79e0 is located in stack of thread T0 at offset 4128 in frame
    #0 0x10710783f in new_VioSSLFd viosslfactories.c:328

  This frame has 2 object(s):
    [32, 4128) 'cipher_list'
    [4256, 4288) '_db_stack_frame_' <== Memory access at offset 4128 partially underflows this variable

**Comment by laurynas-biveinis** [comment]
Added fix for and a testcase for https://bugs.launchpad.net/percona-server/+bug/1596845 / http://bugs.mysql.com/bug.php?id=82026

http://jenkins.percona.com/job/percona-server-5.5-param/1249/

File: vio/viosslfactories.c [hunk 2]
```diff
@@ -238,8 +386,26 @@ new_VioSSLFd(const char *key_file, const char *cert_file,
...
     none of the provided ciphers could be selected
   */
-  if (cipher &&
-      SSL_CTX_set_cipher_list(ssl_fd->ssl_context, cipher) == 0)
+  strcat(cipher_list, tls_cipher_blocked);
+  if (cipher)
+  {
+    if (strlen(cipher_list) + strlen(cipher) + 1 > sizeof(cipher_list))
+    {
+      *error= SSL_INITERR_CIPHERS;
```
Comment (percona-ysorokin) [inline vio/viosslfactories.c:395]: Definitely need
`SSL_CTX_free(ssl_fd->ssl_context);` before `my_free(ssl_fd);`

Probably, add
`DBUG_PRINT("error", ("SSL cipher is too long: %s", sslGetErrString(*error)));`
`report_errors();`
as well

File: vio/viosslfactories.c [hunk 3]
```diff
@@ -238,8 +386,26 @@ new_VioSSLFd(const char *key_file, const char *cert_file,
     NOTE: SSL_CTX_set_cipher_list will return 0 if
     none of the provided ciphers could be selected
   */
-  if (cipher &&
-      SSL_CTX_set_cipher_list(ssl_fd->ssl_context, cipher) == 0)
+  strcat(cipher_list, tls_cipher_blocked);
```
Comment (percona-ysorokin) [inline vio/viosslfactories.c:390]: Similar to the DEBUG_ASSERT below, I would also add
`DBUG_ASSERT(strlen(tls_cipher_blocked) + 1 <= sizeof(cipher_list));` before `strcat()`

File: vio/viosslfactories.c (hunk 2 above)
Comment (laurynas-biveinis) [inline vio/viosslfactories.c:395]: report_errors() is for SSL library errors, which we shouldn't have at this point

Comment (laurynas-biveinis) [inline vio/viosslfactories.c:395]: Likewise for sslGetErrString in DBUG_PRINT

**Comment by laurynas-biveinis** [comment]
http://jenkins.percona.com/job/mysql-5.7-param/228/

**Comment by percona-ysorokin** [comment]
LGTM

**Pull request merged**